src/plugins/data_source/server/plugin.ts

```diff
@@ -32,5 +47,17 @@ export class DataSourcePlugin implements Plugin<DataSourcePluginSetup, DataSourc
     return {};
   }
 
-  public stop() {}
+  public stop() {
+    this.dataSourceClient.close();
+  }
+
+  private createRouteHandlerContext = (
+    core: CoreSetup
+  ): IContextProvider<RequestHandler<unknown, unknown, unknown>, 'data_source'> => {
+    return async (context, req) => {
+      const [{ savedObjects }] = await core.getStartServices();
+      this.dataSourceClient.attachSavedObjectClient(savedObjects.getScopedClient(req));
+      return new DataSourceRouteHandlerContext(this.dataSourceClient, this.logger);
```

Comment:
dummy question, how is DataSourceRouteHandlerContext related to IContextProvider? also is the name mis-matching: ...contextProvider vs ...context

Reply:
We provide our plugin context as a request handler context. `DataSourceRouteHandlerContext` contains the property we want to register. The register function requires to pass in a provider, which you can see is a type of `IContextProvider`

```
registerRouteHandlerContext: <T extends keyof RequestHandlerContext>(
    contextName: T,
    provider: RequestHandlerContextProvider<T>v
  ) => RequestHandlerContextContainer;

export type RequestHandlerContextProvider<
  TContextName extends keyof RequestHandlerContext
> = IContextProvider<RequestHandler<any, any, any>, TContextName>;
```